File: 06-observability/01-three-signals/python/signal_cost.py
"""
Layer 6 Topic 1 - What one unit of telemetry costs the process emitting it.

Why Python: this is the production runtime. Every cost below is paid by the
interpreter, in the request path, on the same thread as the handler. Python is
also the language where the single most common instrumentation bug -- an
argument evaluated for a log line that is never written -- is invisible in
review and expensive at runtime.

What this demonstrates
----------------------
`three_signals.py` shows what each signal can *answer*. This shows what each
signal *costs to emit*, which is the other half of the cardinality/retention
tradeoff and the half that lands in your p99 rather than in your bill.

Five operations, timed on this machine:

  1. counter add      - dict lookup on a bounded label tuple + increment
  2. span record      - object allocation, timestamps, six attributes, append
  3. log line (INFO)  - json.dumps + a real logging.Logger + handler + stream
  4. debug, DISABLED, argument built eagerly    <- the bug
  5. debug, DISABLED, guarded by isEnabledFor   <- the fix

Operations 4 and 5 emit nothing at all. Their cost difference is pure waste,
and it is paid on every request in production forever.

This measures the *shape* of the cost with hand-rolled metric and span stores,
not the OpenTelemetry SDK (which is not installed here). A real SDK adds work
on top of these numbers -- context lookup, attribute validation, view
matching, batching -- it never subtracts any.

What to look for in the output
------------------------------
- The ratio between a counter add and a log line. It is not 2x.
- Rows 4 and 5. Row 4 does exactly as much useful work as row 5: none.
- The last block: telemetry as a percentage of one core at a given request
  rate. That is the number to take to a code review.

Run:  python3 signal_cost.py
"""

import json
import logging
import sys
import time

ITERATIONS = 200_000

# Defeats nothing in CPython (there is no optimizer to defeat here), but kept
# symmetrical with the Rust/C++/Java versions of this file, where it matters.
SINK = [0]


class CounterStore:
    """A bounded-cardinality counter, the way a metrics SDK stores one: a map
    from an ordered label tuple to a number. The cost is the map lookup, which
    is why bounded label sets are cheap and unbounded ones are Topic 4."""

    def __init__(self):
        self.series = {}

    def add(self, labels, value=1):
        self.series[labels] = self.series.get(labels, 0) + value


class Span:
    """__slots__ because a real span is allocated per operation per request and
    a __dict__ per span is measurable at four figures of RPS."""

    __slots__ = ("name", "trace_id", "span_id", "start_ns", "end_ns", "attributes")

    def __init__(self, name, trace_id, span_id, attributes):
        self.name = name
        self.trace_id = trace_id
        self.span_id = span_id
        self.attributes = attributes
        self.start_ns = time.perf_counter_ns()
        self.end_ns = 0

    def end(self):
        self.end_ns = time.perf_counter_ns()


class CountingStream:
    """Stands in for the pipe to your log shipper. Counts bytes so we can also
    report the volume side of the tradeoff, and discards them so we are not
    benchmarking a terminal."""

    def __init__(self):
        self.bytes_written = 0

    def write(self, chunk):
        self.bytes_written += len(chunk)
        return len(chunk)

    def flush(self):
        pass


def build_logger(stream):
    logger = logging.getLogger("signal_cost")
    logger.handlers.clear()
    handler = logging.StreamHandler(stream)
    handler.setFormatter(logging.Formatter("%(message)s"))
    logger.addHandler(handler)
    logger.propagate = False
    # INFO, so every logger.debug() below is disabled -- exactly the production
    # configuration in which the eager-argument bug hides.
    logger.setLevel(logging.INFO)
    return logger


def bench(label, fn, iterations=ITERATIONS):
    fn()  # warm the code path so we are not timing first-call import work
    start = time.perf_counter_ns()
    for _ in range(iterations):
        fn()
    elapsed = time.perf_counter_ns() - start
    return label, elapsed / iterations


def main():
    stream = CountingStream()
    logger = build_logger(stream)
    counter = CounterStore()

    # One request's worth of realistic data: an order being priced.
    order = {
        "order_id": "ord_8f31c2",
        "customer_id": "cus_00194",
        "items": [{"sku": "SKU-1", "qty": 2}, {"sku": "SKU-7", "qty": 1}],
        "discount": 0.15,
        "currency": "GBP",
    }
    labels = (("http.request.method", "GET"),
              ("http.route", "/orders/{id}"),
              ("http.response.status_code", "200"))
    attributes = {
        "http.request.method": "GET",
        "http.route": "/orders/{id}",
        "http.response.status_code": 200,
        "db.system.name": "postgresql",
        "customer.id": order["customer_id"],
        "order.id": order["order_id"],
    }

    def counter_add():
        counter.add(labels)

    def span_record():
        span = Span("GET /orders/{id}", "4bf92f3577b34da6a3ce929d0e0e4736",
                    "00f067aa0ba902b7", attributes)
        span.end()
        SINK[0] += span.end_ns - span.start_ns

    def log_info():
        logger.info(json.dumps({
            "level": "info",
            "msg": "order priced",
            "order_id": order["order_id"],
            "customer_id": order["customer_id"],
            "duration_ms": 12.4,
        }))

    def log_debug_eager():
        # THE BUG. The logger is at INFO, so this line is never written. The
        # json.dumps() still runs, on every request, forever. Nothing in the
        # code review flags it: it reads as "a debug log".
        logger.debug("pricing payload=" + json.dumps(order))

    def log_debug_guarded():
        # THE FIX. One branch. Lift this into any PR where a log argument is
        # anything other than an already-materialised value.
        if logger.isEnabledFor(logging.DEBUG):
            logger.debug("pricing payload=" + json.dumps(order))

    rows = [
        bench("counter.add (3 bounded labels)", counter_add),
        bench("span create + end (6 attrs)", span_record),
        bench("log INFO, one JSON line", log_info),
        bench("log DEBUG (disabled), eager argument", log_debug_eager),
        bench("log DEBUG (disabled), isEnabledFor guard", log_debug_guarded),
    ]

    print("=" * 74)
    print("COST OF EMITTING ONE UNIT OF TELEMETRY   (Python %s, n=%d)"
          % (".".join(map(str, sys.version_info[:3])), ITERATIONS))
    print("=" * 74)
    print("%-42s %12s" % ("operation", "ns/op"))
    for label, ns in rows:
        print("%-42s %12.0f" % (label, ns))

    eager = rows[3][1]
    guarded = rows[4][1]
    print("\nRows 4 and 5 both emit nothing. Row 4 costs %.0f ns more than row 5"
          % (eager - guarded))
    print("for exactly zero output. At 8 disabled debug calls per request and")
    print("1000 req/s that is %.1f ms/s of CPU spent producing nothing."
          % (8 * 1000 * (eager - guarded) / 1e6))

    # A realistic per-request instrumentation budget for a FastAPI handler:
    # one RED counter, three spans (server + db + downstream), two INFO lines,
    # eight disabled debug lines.
    per_request_ns = (rows[0][1] + 3 * rows[1][1] + 2 * rows[2][1]
                      + 8 * rows[3][1])
    per_request_fixed_ns = (rows[0][1] + 3 * rows[1][1] + 2 * rows[2][1]
                            + 8 * rows[4][1])
    print("\nOne request = 1 counter + 3 spans + 2 INFO logs + 8 disabled debug logs")
    print("  as written : %8.1f us/request   -> %.1f%% of one core at 1000 req/s"
          % (per_request_ns / 1000, per_request_ns / 1e9 * 1000 * 100))
    print("  with guards: %8.1f us/request   -> %.1f%% of one core at 1000 req/s"
          % (per_request_fixed_ns / 1000, per_request_fixed_ns / 1e9 * 1000 * 100))
    print("\nBytes written by the INFO logs alone: %d for %d lines (%.0f B/line)."
          % (stream.bytes_written, ITERATIONS + 1,
             stream.bytes_written / (ITERATIONS + 1)))
    print("That is the retention side. The ns/op column is the latency side.")
    print("(sink=%d, printed so nothing above can be optimised away)" % SINK[0])


if __name__ == "__main__":
    main()

File: 06-observability/01-three-signals/nodejs/signal_cost.js
/*
 * Layer 6 Topic 1 - What one unit of telemetry costs the process emitting it.
 *
 * Why Node: it is the strictest version of the bill. Python pays for telemetry
 * on a thread that at least *could* have been one of several; Node pays for it
 * on the one thread that runs every concurrent request in the process. A
 * microsecond spent formatting a log line is a microsecond during which no
 * other request in flight can make progress, so here the ns/op column is not a
 * CPU-budget number, it is a concurrency-capacity number.
 *
 * The second Node-specific trap, and the reason this file writes to a counting
 * sink instead of stdout: when Node's stdout is a TTY or a file it writes
 * *synchronously*. `console.log` in a request handler is a blocking write on
 * the event loop thread. Piped stdout is async. The same code therefore has
 * different latency depending on how the process was started, which is why
 * production Node logs through a buffered transport and never through console.
 *
 * What this demonstrates
 * ----------------------
 * Five operations, timed on this machine:
 *   1. counter add      - Map lookup on a bounded label key + increment
 *   2. span record      - object allocation, hrtime timestamps, six attributes
 *   3. log line (INFO)  - JSON.stringify + write to the sink
 *   4. debug, DISABLED, template literal built eagerly    <- the bug
 *   5. debug, DISABLED, guarded by a level check          <- the fix
 *
 * Operations 4 and 5 emit nothing. The difference between them is pure waste,
 * paid per request, forever.
 *
 * This measures the shape of the cost with a hand-rolled metric/span/log
 * store, not the OpenTelemetry SDK (not installed here). A real SDK adds work
 * on top of these numbers; it never subtracts any.
 *
 * What to look for in the output
 * ------------------------------
 * - The counter-to-log-line ratio. It is not 2x.
 * - Row 4 versus row 5: identical output (none), very different cost.
 * - The warm-up line. V8 tiers up; the first thousand iterations of anything
 *   are interpreted, so a benchmark without warm-up measures the compiler.
 *
 * Run:  node signal_cost.js
 */

'use strict';

const ITERATIONS = 200_000;
const WARMUP = 20_000;

// Printed at the end so V8 cannot decide the work below is dead.
let sink = 0n;

class CounterStore {
  constructor() {
    this.series = new Map();
  }
  add(key, value = 1) {
    this.series.set(key, (this.series.get(key) || 0) + value);
  }
}

class Span {
  constructor(name, traceId, spanId, attributes) {
    this.name = name;
    this.traceId = traceId;
    this.spanId = spanId;
    this.attributes = attributes;
    this.startNs = process.hrtime.bigint();
    this.endNs = 0n;
  }
  end() {
    this.endNs = process.hrtime.bigint();
  }
}

// Stands in for the pipe to the log shipper: counts bytes, discards them.
class CountingSink {
  constructor() {
    this.bytesWritten = 0;
    this.lines = 0;
  }
  write(line) {
    this.bytesWritten += Buffer.byteLength(line);
    this.lines += 1;
  }
}

// A logger of the shape every Node logging library has: a numeric level, and
// methods that check it. pino and winston both look like this underneath.
const LEVELS = { debug: 20, info: 30, warn: 40, error: 50 };

class Logger {
  constructor(sinkStream, level) {
    this.sink = sinkStream;
    this.level = LEVELS[level];
  }
  isEnabled(level) {
    return LEVELS[level] >= this.level;
  }
  debug(line) {
    if (LEVELS.debug >= this.level) this.sink.write(line + '\n');
  }
  info(line) {
    if (LEVELS.info >= this.level) this.sink.write(line + '\n');
  }
}

function bench(label, fn) {
  for (let i = 0; i < WARMUP; i++) fn(); // let V8 tier up before timing
  const start = process.hrtime.bigint();
  for (let i = 0; i < ITERATIONS; i++) fn();
  const elapsed = process.hrtime.bigint() - start;
  return { label, nsPerOp: Number(elapsed) / ITERATIONS };
}

function main() {
  const sinkStream = new CountingSink();
  const logger = new Logger(sinkStream, 'info'); // debug is disabled
  const counter = new CounterStore();

  const order = {
    order_id: 'ord_8f31c2',
    customer_id: 'cus_00194',
    items: [{ sku: 'SKU-1', qty: 2 }, { sku: 'SKU-7', qty: 1 }],
    discount: 0.15,
    currency: 'GBP',
  };
  const labelKey = 'GET|/orders/{id}|200';
  const attributes = {
    'http.request.method': 'GET',
    'http.route': '/orders/{id}',
    'http.response.status_code': 200,
    'db.system.name': 'postgresql',
    'customer.id': order.customer_id,
    'order.id': order.order_id,
  };

  const ops = [
    bench('counter.add (3 bounded labels)', () => {
      counter.add(labelKey);
    }),
    bench('span create + end (6 attrs)', () => {
      const span = new Span('GET /orders/{id}', '4bf92f3577b34da6a3ce929d0e0e4736',
        '00f067aa0ba902b7', attributes);
      span.end();
      sink += span.endNs - span.startNs;
    }),
    bench('log INFO, one JSON line', () => {
      logger.info(JSON.stringify({
        level: 'info',
        msg: 'order priced',
        order_id: order.order_id,
        customer_id: order.customer_id,
        duration_ms: 12.4,
      }));
    }),
    bench('log DEBUG (disabled), eager template literal', () => {
      // THE BUG. logger.level is info, so nothing is written -- but the
      // template literal, and therefore JSON.stringify(order), is evaluated
      // before `debug` is even called. Arguments are evaluated at the call
      // site in JavaScript; there is no lazy form of a template literal.
      logger.debug(`pricing payload=${JSON.stringify(order)}`);
    }),
    bench('log DEBUG (disabled), level-check guard', () => {
      // THE FIX. Same shape as pino's `logger.isLevelEnabled('debug')`.
      if (logger.isEnabled('debug')) {
        logger.debug(`pricing payload=${JSON.stringify(order)}`);
      }
    }),
  ];

  const line = '='.repeat(74);
  console.log(line);
  console.log(`COST OF EMITTING ONE UNIT OF TELEMETRY   (Node ${process.version}, n=${ITERATIONS})`);
  console.log(line);
  console.log('operation'.padEnd(46) + 'ns/op'.padStart(12));
  for (const op of ops) {
    console.log(op.label.padEnd(46) + op.nsPerOp.toFixed(0).padStart(12));
  }

  const eager = ops[3].nsPerOp;
  const guarded = ops[4].nsPerOp;
  console.log(`\nRows 4 and 5 both emit nothing. Row 4 costs ${(eager - guarded).toFixed(0)} ns more`);
  console.log('than row 5 for exactly zero output. At 8 disabled debug calls per');
  console.log(`request and 1000 req/s that is ${((8 * 1000 * (eager - guarded)) / 1e6).toFixed(1)} ms/s of event-loop time`);
  console.log('spent producing nothing -- and event-loop time is the resource every');
  console.log('other in-flight request is waiting for.');

  const perRequest = ops[0].nsPerOp + 3 * ops[1].nsPerOp + 2 * ops[2].nsPerOp + 8 * ops[3].nsPerOp;
  const perRequestFixed = ops[0].nsPerOp + 3 * ops[1].nsPerOp + 2 * ops[2].nsPerOp + 8 * ops[4].nsPerOp;
  console.log('\nOne request = 1 counter + 3 spans + 2 INFO logs + 8 disabled debug logs');
  console.log(`  as written : ${(perRequest / 1000).toFixed(1)} us of event loop per request`);
  console.log(`  with guards: ${(perRequestFixed / 1000).toFixed(1)} us of event loop per request`);
  console.log(`  ceiling on requests/sec from telemetry alone, as written: ${(1e9 / perRequest).toFixed(0)}`);
  console.log('  (that ceiling assumes the handler itself costs nothing, which is the');
  console.log('   point -- telemetry sets a hard cap on a single-threaded runtime.)');

  console.log(`\nBytes written by the INFO logs: ${sinkStream.bytesWritten} over ${sinkStream.lines} lines`
    + ` (${(sinkStream.bytesWritten / sinkStream.lines).toFixed(0)} B/line).`);
  console.log(`(sink=${sink}, printed so nothing above can be optimised away)`);
}

main();
